fvm/environment/unsafe_random_generator_test.go

```diff
@@ -1,36 +1,64 @@
 package environment_test
 
 import (
+	"fmt"
+	"math"
+	mrand "math/rand"
 	"testing"
 
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"gonum.org/v1/gonum/stat"
 
 	"github.com/onflow/flow-go/fvm/environment"
 	"github.com/onflow/flow-go/fvm/tracing"
 	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/utils/unittest"
 )
 
 func TestUnsafeRandomGenerator(t *testing.T) {
-	t.Run("UnsafeRandom doesnt re-seed the random", func(t *testing.T) {
-		bh := &flow.Header{}
-
+	// basic randomness test to check outputs are "uniformly" spread over the
+	// output space
+	t.Run("randomness test", func(t *testing.T) {
+		bh := unittest.BlockHeaderFixtureOnChain(flow.Mainnet.Chain().ChainID())
 		urg := environment.NewUnsafeRandomGenerator(tracing.NewTracerSpan(), bh)
 
-		// 10 random numbers. extremely unlikely to get the same number all the time and just fail the test by chance
-		N := 10
-
-		numbers := make([]uint64, N)
+		sampleSize := 80000
+		tolerance := 0.05
+		n := 10 + mrand.Intn(100)
+		distribution := make([]float64, n)
 
-		for i := 0; i < N; i++ {
-			u, err := urg.UnsafeRandom()
+		// partition all outputs into `n` classes and compute the distribution
+		// over the partition. Each class is `classWidth`-big
+		classWidth := math.MaxUint64 / uint64(n)
+		// populate the distribution
+		for i := 0; i < sampleSize; i++ {
+			r, err := urg.UnsafeRandom()
 			require.NoError(t, err)
-			numbers[i] = u
+			distribution[r/classWidth] += 1.0
 		}
+		stdev := stat.StdDev(distribution, nil)
+		mean := stat.Mean(distribution, nil)
+		assert.Greater(t, tolerance*mean, stdev, fmt.Sprintf("basic randomness test failed. stdev %v, mean %v", stdev, mean))
```

Comment:
this check seems weird.  std dev is meaningless for uniform distribution.  you want something like

```
expectedMean := math.MaxUint64 / 2
expectedMean * (1 - tolerance) <= mean <= expectedMean * (1 + tolerance)
```

instead?

Reply:
I have two comments to your suggestion:

- Imagine a dice with the possible outputs `[1,2,3,4,5,6]` and the random variable being the value we read after throwing the dice. In this case the expected mean according to your formula is `3,5`.  The uniform distribution we would like a perfect dice to have is `[1/6, 1/6, 1/6, 1/6, 1/6, 1/6]`. How can we verify that with a practical statistical test?  Let's say we throw the dice 100 times, and we get 50 times `1`, and 50 times `6`.  The mean in this case is exactly the expected mean `3,5` (  (1*50+6*50)/100 ), and we pass the inequality above with any small tolerance we chose. However the dice we tested is clearly biased and its distribution is far from uniform. 

- The test of the PR takes a different approach. It throws the dice 100 times and computes the frequency of each possible output in `{1,2,3,4,5,6}`. If the dice is perfect, we would get an array approximately `[16 ,16, 16, 16, 16 ,16]`. In other terms, the expected mean is `100/6 = 16,66` and we would like to test that each value of the array is close to that expected mean. The standard deviation is a basic statistical tool to measure that: it quantifies the dispersion of many values compared to their mean. A low standard deviation means that all values are close to the mean, a high standard deviation means the values are spread out from the mean. The biased dice in the example above gives `[50, 0, 0, 0, 0, 50]` and it gives a very large standard deviation. This test remains a basic statistical test that aims to find obvious issues or software bugs, it's relatively quick so that it can run as a unit test. But of course there are deeper randomness statistical tests that are beyond the scope of a unit test.

Reply:
Does it make sense to plug this into some existing test suite for random generators? like https://github.com/terrillmoore/NIST-Statistical-Test-Suite in the future?

Reply:
Those are an example of the deep statistical tests that I mentioned above. They are heavy (take time and write in lots of memory) and are used to test a TRG (true random generator) or a new implementation of a PRG. In our case we are using a known and established PRG underneath, called ChaCha20. Such PRG has been already studied on a cryptanalysis level (theory level) and statistical level (running tests on its outputs). Our code repo tests do not mean to re-evaluate ChaCha20, they only try to detect biases resulting from implementation bugs, while running relatively fast. 

Btw, the crypto lib (where ChaCha20 is implemented) already checks that the implementation is compliant to https://www.rfc-editor.org/rfc/rfc7539, so we don't have to repeat it here.